File: Chapter 03/JS/section_3_3_5_dip/before/coach.js
const Forward = require('./forward');
const Midfielder = require('./midfielder');

/**
 * ANTI-PATTERN: Violation of the Dependency Inversion Principle (DIP).
 * * ARCHITECTURE PROBLEM: High-level modules (Coach) should not depend on 
 * low-level modules (Forward/Midfielder). Both should depend on abstractions.
 * * By using 'require' and 'new' inside this class, the Coach is now 
 * "hard-wired" to these specific players. 
 * * WHY THIS FAILS:
 * - Rigid: You cannot swap a Forward for a Winger without editing the Coach file.
 * - Untestable: You cannot unit test the Coach without also creating real 
 * instances of the players. You are stuck with whatever is in those files.
 */
class Coach {
    constructor() {
        // 🚨 ARCHITECTURE WARNING: The 'new' keyword is glue. 
        // The Coach is taking on the responsibility of "manufacturing" its 
        // own dependencies instead of just using them.
        this.forward = new Forward();
        this.midfielder = new Midfielder();
    }

    executeGamePlan() {
        this.forward.attack();
        this.midfielder.controlMidfield();
    }
}

module.exports = Coach;

File: Chapter 03/Python/section_3_3_5_dip/before/coach.py
from .forward import Forward
from .midfielder import Midfielder

"""
ANTI-PATTERN: Violation of the Dependency Inversion Principle (DIP).

ARCHITECTURE PROBLEM: High-level modules (Coach) should not depend on 
low-level modules (Forward/Midfielder). Both should depend on abstractions.

By using 'import' and 'self.player = Player()' inside this class, the Coach 
is now "hard-wired" to these specific implementations.

WHY THIS IS BRITTLE:
- Rigidity: If you want to swap a Midfielder for a Winger, you have to 
  manually edit the Coach class.
- Untestable: You cannot unit test the Coach in isolation. You are forced 
  to create real instances of the players every time.
- Violation of "New is Glue": The Coach is responsible for creating its 
  own team instead of just leading the team it's given.
"""

class Coach:
    def __init__(self):
        # 🚨 ARCHITECTURE WARNING: Direct Instantiation.
        # The Coach is now stuck with these specific classes forever.
        self.forward = Forward()
        self.midfielder = Midfielder()

    def execute_game_plan(self):
        # 🚨 ARCHITECTURE WARNING: Specific Method Coupling.
        # The Coach has to know the exact method names (attack, control_midfield)
        # for each specific player type.
        self.forward.attack()
        self.midfielder.control_midfield()